src/program.rs

```diff
@@ -401,77 +405,209 @@ impl From<ColorOpt> for ColorChoice {
     }
 }
 
+#[cfg(feature = "doc")]
+pub use doc::DocFormat;
+
 #[cfg(feature = "doc")]
 mod doc {
     use crate::cache::Cache;
-    use crate::error::{Error, IOError};
+    use crate::error::{Error, IOError, SerializationError};
     use crate::term::{RichTerm, Term};
     use codespan::FileId;
     use comrak::arena_tree::NodeEdge;
     use comrak::nodes::{Ast, AstNode, NodeCode, NodeHeading, NodeValue};
     use comrak::{format_commonmark, parse_document, Arena, ComrakOptions};
+    use serde::{Deserialize, Serialize};
+    use std::collections::HashMap;
+    use std::fmt;
     use std::io::Write;
 
-    /// Create a markdown file with documentation for the specified FileId.
+    /// Create a markdown or json file with documentation for the specified FileId.
     pub fn output_doc(
```

Comment:
any reason not to make it be an `output()` method of `ExtractedDoc`, rather than a freestanding function? Also, it's a matter a taste, but I have a slight preference toward `output_json` and `output_markdown` as an interface (or even `write_json` or `write_markdown`, as it is customary with serde modules), rather than one function parametrized by (something isomorphic to) a boolean.

Reply:
The `to_json` and `to_markdown` should probably really be called `write_json` and `write_markdown`. I think those would be the `output_xxx` functions you're thinking of. The `output_doc` function might actually make more sense as a function on `DocFormat`, since it takes care of extracting the documentation from a file.

Reply:
To me, `DocFormat` is really just an basic enum, so that would be weird to implement the logic on this one. My initial idea was rather to have a function `ExtractedDocumentation::extract_from_file` or whatever you call it, that would take the same arguments (excepted format), and produce an `ExtractedDocumentation` (because that's what you do in the middle, right). Then, separate functions to write this documentation in whatever format you want to a stream, which would be `write_json` and `write_markdown` (or, if you really fancy it, one `write` function taking a `Format` argument). IMHO it makes more sense that a function traversing `ExtractedDocumentation` to write to a stream is attached to `ExtractedDocumentation`. Here it seems to be an artifact from the fact that `output_doc` is actually the composition of morally smaller functions.

Reply:
Agreed for `to_json` vs `write_json`, I was quoting the name from serde, but `to_xxx` generally produces a `String`, and `write_xxx` write to a stream.

Reply:
I agree with the second option as well. This is also more in line with how we already implement the `export` subcommand, I think.

> Separately, I feel like renaming SerializationError to ExportError and adding an enum case for MissingDocumentation to it. At the moment, Nickel just doesn't write anything to the specified file if it can't find any documentation to extract.

Sounds sensible, yes :+1: